### DISCUSSION TITLE
Torments of deployment.

### DISCUSSION THREAD

**Discussion by JHerwig** [nuxt/nuxt#9711] [Questions (Nuxt 2)] [answered]
Hey folks,
I wish to auto-deploy my universal nuxt app. 
What I did: I configured git, that on push on release it copies all files into a specific folder. 

The next step is to execute a script on each push. This script should kill the current node process, build the project, and (re-)start the application. I imagined something like:
```bash
$ pkill -f <projectname>
$ npm run build
$ npm run start
```

But I already struggling to deploy the app by hand. I encounter these problems:
1. Problem: Wrong version?
I installed npm via `apt-get`. But while running `build` I got
```bash
 WARN  You are using an unsupported version of Node.js (v10.19.0). It is recommended to use the latest LTS version (https://nodejs.org/en/about/releases)
```
Trying ` $ apt-get install --only-upgrade` didn't help either. Following a stackoverflow I tried:
```bash
$ npm cache clean -f
$ npm install -g n
$ sudo n stable
```
That solved my problem, but I don't know why that is necessary. Am I doing something wrong?

2. Problem: Many errors, .nuxtignore gets ignored?
I have some test files which are mentioned in .nuxtignore. Locally they are ignored correctly but on the server side, the build keeps failing because of them. 
As a workaround, I deleted them and put them into .gitignore. But can anyone tell me, why .nuxtignore is not working?

3. Problem: Not enough permission?
While `npm run build` i got:
```bash
gyp WARN EACCES user "root" does not have permission to access the dev dir "/root/.node-gyp/14.17.5"
gyp WARN EACCES attempting to reinstall using temporary dev dir "/var/www/node/websdv_frontend/node_modules/node-sass/.            
```
Can anyone tell me what to do?

I am really frustrated. Deploying an application that works locally shouldn't be so complicated.

Since we're at deployment: Is there a smart way to have 2 .env for production and development without using gitingore?

**Comment by TheAlexLichter** [comment, marked as answer]
1. Don't install Node.js via the package manager. It will almost always be an old version. Instead, follow the advice on the official Node.js website and use a node version manager like `n` or `nvm` ☺️ 
2. Hard to guess. Possibly because of `\` vs. `/` (windows vs. unix paths) in your `.nuxtignore`?
3. You shouldn't install NPM so only root can use it. For a quick workaround, see https://github.com/nodejs/node-gyp/issues/454. Besides that, I'd suggest using sth. like `Docker` to make deployment easier

**Comment by JHerwig** [comment]
I got some help from friends and the Discord Nuxt Community. 
Here is my current state: 
1. Solved using `n`
2.  Not solved, I've added all "test" files to gitignore.
3. Solved, that was (obvious) a permission problem. I've changed the Group and Owner for all files to a (new created) "node" user. This user can now build and start everything without problems.

But sadly I stuck at the next Problem `Content-Security-Policy`. But I will start another thread about that.
